vm/systemSmartContracts/esdt.go

```diff
@@ -0,0 +1,279 @@
+const canFreeze = "canFreeze"
+const canWipe = "canWipe"
+
+const conversionBase = 10
+
+//TODO: think about how should we enable issuing of shorter names - maybe only a special address can do it
+// a little bit centralized but it resolves the problem of issuing commonly known token names
+
+type esdt struct {
+	eei             vm.SystemEI
+	gasCost         vm.GasCost
+	baseIssuingCost *big.Int
+	ownerAddress    []byte
+	eSDTSCAddress   []byte
+	marshalizer     marshal.Marshalizer
+	hasher          hashing.Hasher
+}
+
+// ArgsNewESDTSmartContract defines the arguments needed for the esdt contract
+type ArgsNewESDTSmartContract struct {
+	Eei           vm.SystemEI
+	GasCost       vm.GasCost
+	ESDTSCConfig  config.ESDTSystemSCConfig
+	ESDTSCAddress []byte
+	Marshalizer   marshal.Marshalizer
+	Hasher        hashing.Hasher
+}
+
+// NewESDTSmartContract create the esdt smart contract, which controls the issuing of tokens
+func NewESDTSmartContract(args ArgsNewESDTSmartContract) (*esdt, error) {
+	if check.IfNil(args.Eei) {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, vm.ErrNilMarshalizer
+	}
+	if check.IfNil(args.Hasher) {
+		return nil, vm.ErrNilHasher
+	}
+
+	baseIssuingCost, ok := big.NewInt(0).SetString(args.ESDTSCConfig.BaseIssuingCost, conversionBase)
+	if !ok || baseIssuingCost.Cmp(big.NewInt(0)) < 0 {
+		return nil, vm.ErrInvalidBaseIssuingCost
+	}
+
+	return &esdt{
+		eei:             args.Eei,
+		gasCost:         args.GasCost,
+		baseIssuingCost: baseIssuingCost,
+		ownerAddress:    []byte(args.ESDTSCConfig.OwnerAddress),
+		eSDTSCAddress:   args.ESDTSCAddress,
+		hasher:          args.Hasher,
+		marshalizer:     args.Marshalizer,
+	}, nil
+}
+
+// Execute  calls one of the functions from the esdt smart contract and runs the code according to the input
+func (e *esdt) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if CheckIfNil(args) != nil {
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case "_init":
+		return e.init(args)
+	case "issue":
+		return e.issue(args)
+	case "issueProtected":
+		return e.issueProtected(args)
+	case "burn":
+		return e.burn(args)
+	case "mint":
+		return e.mint(args)
+	case "freeze":
+		return e.freeze(args)
+	case "wipe":
+		return e.wipe(args)
+	case "pause":
+		return e.pause(args)
+	case "unPause":
+		return e.unpause(args)
+	case "claim":
+		return e.claim(args)
+	case "configChange":
+		return e.configChange(args)
+	case "esdtControlChanges":
+		return e.esdtControlChanges(args)
+	}
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) init(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	scConfig := &ESDTConfig{
+		OwnerAddress:       e.ownerAddress,
+		BaseIssuingCost:    e.baseIssuingCost,
+		MinTokenNameLength: minLengthForTokenName,
+		MaxTokenNameLength: maxLengthForTokenName,
+	}
+	marshalledData, err := e.marshalizer.Marshal(scConfig)
+	log.LogIfError(err, "marshall error on esdt init function")
```

Comment:
if err != nil should not return with error?

Reply:
it is not critical.